cooperator/models/subscription_request.py

```diff
@@ -103,51 +101,45 @@ def _find_partner_from_create_vals(self, vals):
-            ]
-            # we don't use partner.coop_candidate because we want to also
-            # handle draft and waiting requests.
-            if member or self.search(pending_requests_domain):
-                vals["type"] = "increase"
-            if member:
-                vals["already_cooperator"] = True
-            if not cooperative_membership:
-                cooperative_membership = partner.create_cooperative_membership(
-                    company_id
-                )
-            elif not cooperative_membership.cooperator:
-                cooperative_membership.cooperator = True
-
-        subscription_request = super().create(vals)
-        # TODO: This should probably not be in the create method. There may need
-        # to be a stage after draft in which this e-mail is sent, or the e-mail
-        # should exclusively be sent from `cooperator_website`. See #73 for
-        # some comments, and for a reverted implementation of the extra state.
-        subscription_request._send_confirmation_mail()
-        return subscription_request
-
-    @api.model
-    def create_comp_sub_req(self, vals):
-        warnings.warn(
-            "subscription.request.create_comp_sub_req() is deprecated. "
-            "please use .create() instead.",
-            DeprecationWarning,
-        )
-        return self.create(vals)
+    def _adapt_create_vals_and_membership_from_partner(self, vals, partner):
+        """
+        Check for existing cooperative membership for the partner, create or
+        update it if needed and set vals accordingly.
+        """
+        company_id = vals.get("company_id", self.env.company.id)
+        cooperative_membership = partner.get_cooperative_membership(company_id)
+        member = cooperative_membership and cooperative_membership.member
+        pending_requests_domain = [
+            ("company_id", "=", company_id),
+            ("partner_id", "=", partner.id),
+            ("state", "in", ("draft", "waiting", "done")),
+        ]
+        # we don't use partner.coop_candidate because we want to also
+        # handle draft and waiting requests.
+        if member or self.search(pending_requests_domain):
+            vals["type"] = "increase"
+        if member:
+            vals["already_cooperator"] = True
+        if not cooperative_membership:
+            cooperative_membership = partner.create_cooperative_membership(company_id)
+        elif not cooperative_membership.cooperator:
+            cooperative_membership.cooperator = True
 
-    def check_empty_string(self, value):
-        if value is None or value is False or value == "":
-            return False
-        return True
+    @api.model_create_multi
+    def create(self, vals_list):
+        records = self.browse()
+        for vals in vals_list:
+            partner = self._find_partner_from_create_vals(vals)
+            if partner:
+                self._adapt_create_vals_and_membership_from_partner(vals, partner)
```

Comment:
Amazing, thank you.

Reply:
you’re welcome (it was itching too much). ☺️